### ISSUE TITLE
file.recurse causes 'grain cache flashbacks' - evokes outdated grain values in the cache

### ISSUE THREAD

**Issue by markuskramerIgitt** [saltstack/salt#25499]
I found a reproducable test case for a disturbing problem.

The minion MINION has currently IP-address 10.144.76.86.
In the past it probably has been 10.144.73.58, but not now.

On MINION, I set a grain called magni_config to the value of 3.
In the past it probably has been 1.

To my great surprise file.recurse overwrites the cache of grains and pillars (/var/cache/salt/master/minions/MINION/data.p) with both, outdated values.
As you can see below, it reverts value magni_config to the outdated value of 1 and it wrongly adds the outdated IP-address.
When I look directly into file  /salt/conf/grains on MINION it contains the correct value, proving that the error is 'local' to the salt-master.

Below is the complete and (apart from the minion name) unmodified sequence of action.

I have about 40 minions and about 50% suffer from 'grain cache flashbacks'.
The minion which never executes ‘file.recurse’ does not suffer under 'grain cache flashbacks'.

The documentation says that grains should be 'static'.
What is the precise definition of 'static'?
My minions are Laptops and can change their IP-address, possibly several times a day.
Does this make the IP-address 'not static enough'?

An attempt to explain what I observe:
- The module 'file' keeps an ‘in-memory’ cache of grains.
- This in-memory cache is not invalidated if the module 'grains' updates the file cache.
- When the module 'file' is executed, it overwrites the file cache with the in-memory cache.
- Desaster.

Could you please verify/falsify this hypothesis?

I looked for similar issues and just found:
In https://github.com/saltstack/salt/issues/14691  @ryepup  says:  "It almost seems like state.sls is somehow clearing the custom grains that didn't exist when the minion first booted up."

In https://github.com/saltstack/salt/issues/14691   the hypothesis is that the Windows threading model is related, but here the threading model of the minion can hardly be part of the problem because the minion has the correct grain (in /salt/conf/grains).

```

markus@debian:~$ sudo salt MINION state.sls config
MINION:
----------
          ID: magni_config_setter
    Function: grains.present
        Name: magni_config
      Result: True
     Comment: Set grain magni_config to 3
     Started: 22:53:09.346000
    Duration: 3229.0 ms
     Changes:
              ----------
              magni_config:
                  3

Summary
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1


markus@debian:~$ sudo salt MINION grains.item ipv4
MINION:
    ----------
    ipv4:
        - 10.144.76.86


markus@debian:~$ sudo salt MINION state.sls santa
MINION:
----------
          ID: /apps/dhlcopy/north
    Function: file.recurse
      Result: True
     Comment: The directory /apps/dhlcopy/north is in the correct state
     Started: 22:54:36.113000
    Duration: 811.0 ms
     Changes:

Summary
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1


markus@debian:~$ sudo salt MINION grains.item ipv4
MINION:
    ----------
    ipv4:
        - 10.144.73.58
        - 10.144.76.86


markus@debian:~$ sudo salt MINION grains.item magni_config
MINION:
    ----------
    magni_config:
        1


markus@debian:~ $ sudo salt MINION cmd.run cwd=/salt/conf 'type grains'
MINION:
    magni_config: 3


markus@debian:~$ sudo salt MINION cmd.run cwd=/salt 'salt-call --versions-report'
MINION:
               Salt: 2014.7.2
             Python: 2.7.5 (default, May 15 2013, 22:43:36) [MSC v.1500 32 bit (Intel)]
             Jinja2: 2.7.1
           M2Crypto: 0.21.1
     msgpack-python: 0.4.2
       msgpack-pure: Not Installed
           pycrypto: 2.6
            libnacl: Not Installed
             PyYAML: 3.11
              ioflo: Not Installed
              PyZMQ: 14.1.1
               RAET: Not Installed
                ZMQ: 4.0.4
               Mako: Not Installed


markus@debian:~$ salt --versions-report
                  Salt: 2015.5.0
                Python: 2.7.3 (default, Mar 13 2014, 11:03:55)
                Jinja2: 2.6
              M2Crypto: 0.21.1
        msgpack-python: 0.1.10
          msgpack-pure: Not Installed
              pycrypto: 2.6
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 13.1.0
                  RAET: Not Installed
                   ZMQ: 3.2.3
                  Mako: 0.7.0
 Debian source package: 2015.5.0+ds-1~bpo70+1

```

**Comment by jfindlay** [comment]
@markuskramerIgitt, thanks for the report.

**Comment by jfindlay** [comment]
@markuskramerIgitt, can you confirm whether this still happens on the latest master and minion (2015.5.2 or 2015.5.3)?  Thanks.

**Comment by markuskramerIgitt** [comment]
@jfindlay, I will update the master and minions, probably next week.
In the meantime I found that I reproducably toggle between "current" and "flashback" grains by 

```
saltutil.sync_grains --> "current" grains
file.recurse --> "flashback" grains.
```


If see the grains that "change" with ```diff -y```

```
grains:                                                         grains:
  fqdn_ip4: [192.168.178.48, 2.148.66.70]                     |   fqdn_ip4: [2.148.66.70, 192.168.178.34]
  ip4_interfaces:                                                 ip4_interfaces:
    Cisco Systems VPN Adapter: [2.148.66.70]                        Cisco Systems VPN Adapter: [2.148.66.70]
    Intel(R) Centrino(R) Advanced-N 6205: [192.168.178.48]    |     Intel(R) Centrino(R) Advanced-N 6205: [192.168.178.34]
  ip_interfaces:                                                  ip_interfaces:
    Cisco Systems VPN Adapter: [2.148.66.70]                        Cisco Systems VPN Adapter: [2.148.66.70]
    Intel(R) Centrino(R) Advanced-N 6205: [192.168.178.48]    |     Intel(R) Centrino(R) Advanced-N 6205: [192.168.178.34]
  ipv4: [192.168.178.48, 2.148.66.70]                         |   ipv4: [192.168.178.34, 2.148.66.70]
```



I use this skript to create "snapshots" of the grains
```
#!/usr/bin/env python 
'''
PURPOSE
  Take a readable snapshot of all minion caches.

USAGE
   sudo util/salt-cache.py [OUTPUT_DIR]
	 Whithout OUTPUT_DIR, std-out is used

BACKGROUND
  The salt-master has a cache of the minions grains and pillars

'''

import sys, os
import msgpack, yaml

IS_STDOUT  = len(sys.argv) == 1
OUTPUT_DIR = None
CACHE_DIR  = '/var/cache/salt/master/minions/'

if not IS_STDOUT:
	OUTPUT_DIR = sys.argv[1]
	if not os.path.isdir(OUTPUT_DIR) or os.listdir(OUTPUT_DIR) != []:
		raise SystemExit("The output directory must be empty")

for one_minion in os.listdir (CACHE_DIR):
	in_NAME  = os.path.join (CACHE_DIR, one_minion, "data.p")
	in_STAT  = os.stat(in_NAME)
	yamldump = None
	with open(in_NAME, 'rb') as in_FILEHANDLE:
		yamldump = yaml.dump(msgpack.loads(in_FILEHANDLE.read(), use_list=True))
	if IS_STDOUT:
		print "******************************************"
		print "************** cache of minion " + one_minion
		print yamldump
	else:
		out_NAME = os.path.join(OUTPUT_DIR, one_minion + ".yaml")
		with open(out_NAME, 'w+') as out_FILEHANDLE:
			out_FILEHANDLE.write(yamldump)
		os.utime(out_NAME, (in_STAT.st_atime, in_STAT.st_mtime)) # preserve modification time 

```

**Comment by basepi** [comment]
Does `salt '*' saltutil.sync_grains refresh=True` stop it from flipflopping?

**Comment by markuskramerIgitt** [comment]
```saltutil.sync_grains```  only corrected the grains, so the "flipflopping" was good and necessary.
The problem was the flipflopping of  ```file recurse```

I no longer use ```file recurse```, so it would have to setup a test scenario (not before September).

**Comment by basepi** [comment]
Right, I was wondering if the refresh=True would stop file.recurse from resetting it to the old grains.

A test scenario would be great, otherwise we'll look into this when we have some time. Thanks!

**Comment by markuskramerIgitt** [comment]
Sorry, I did not get it the first time!
I will try out ```refresh=True``` (very probably only after my refresh, err, holidays) and let you know.

**Comment by cachedout** [comment]
@markuskramerIgitt Sorry for the delay in getting back to you on this one. Were you able to try the test that @basepi recommended?

**Comment by basepi** [comment]
@markuskramerIgitt Did you ever get a chance to test this? Since you're the only one running into this I anticipate it being difficult to reproduce, so any additional details we can get would be great.

**Comment by markuskramerIgitt** [comment]
@basepi, I am working on grains now and have will use my documentation on the issue to try to reproduce it.

**Comment by markuskramerIgitt** [comment]
Hi @basepi,
these are the 5 commands and the sls file to reproduce the effect:

```

sudo salt '*' grains.setval testflashback v5 --out=txt
sudo salt '*' state.sls testflashback --state-output=terse --out=txt
sudo salt '*' grains.get testflashback --out=txt
sudo salt '*' saltutil.sync_grains
sudo salt '*' saltutil.sync_grains refresh=True

------------------- /srv/salt/testflashback.sls ---------
test grain flashbacks:
  file.managed:
    - name: /windows/temp/testflashback.txt
    - source: salt://testflashback.txt
------------------- /srv/salt/testflashback.sls ---------

```

**Observations:**
Immediately after `grains.set`, all minions respond to `grains.get` with the correct value (without flashback).
After `file.managed`, only ~ 33% of the minions respond to `grains.get`, ~ 20% of which with a flashback.
When `grains.get` is repeated (~5 min interval), more minions respond, and 1 minion even changes from flashback to correct answer.
All minions repond to `sync_grains`.
After `sync_grains`, all minion respond to `grains.get` without flashbacks.
After `file.manage` (again), the problems toggles back.
There is no difference in `saltutil.sync_grains refresh=True` and `saltutil.sync_grains`, the toggling remains.
It does not matter that the managed file stays the same all of the time.

**Summary:**
I reproducably toggle between "100% and current" (sync_grains) and "33% and 20% flashback" (file.manage) grains
My workaround will be to sync_grains after each file.manage

**Details:**

```

Using testflashback value v5:
42 grains.set
42 file.managed
25 grains.get --> 5 flashback v3

Using testflashback value v6: 
42 grains.set
42 file.managed
25 grains.get --> 5 flashback v3 (same minions as for v5)
26 grains.get --> 4 flashback v3 (1 flashback minion corrected itself to v6)
29 grains.get --> 4 flashback v3 
29 grains.get --> 4 flashback v3 

Using testflashback value v7: 
41 grains.set
41 grains.get --> 0 flashback 
41 file.managed
25 grains.get --> 5 flashback v3 (including the mininon that corrected himself for v6)
27 grains.get --> 4 flashback v3 
40 sync_grains
40 grains.get --> 0 flashback 

```

**Comment by markuskramerIgitt** [comment]
```
 salt --versions-report
                  Salt: 2015.5.3
                Python: 2.7.3 (default, Mar 13 2014, 11:03:55)
                Jinja2: 2.6
              M2Crypto: 0.21.1
        msgpack-python: 0.1.10
          msgpack-pure: Not Installed
              pycrypto: 2.6
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 13.1.0
                  RAET: Not Installed
                   ZMQ: 3.2.3
                  Mako: 0.7.0
               Tornado: Not Installed
 Debian source package: 2015.5.3+ds-1~bpo70+2

```

Minions are a mixture of 2014.7.2 to 2015.5.3.

Flashbacks occured on 2014.7.2  2015.5.0.

I try to upgrade all to 2015.5.3

**Comment by basepi** [comment]
Thanks for the reproduction steps. This issue still mystifies me but maybe if we increase the sample size we can reproduce.

**Comment by basepi** [comment]
(In the meantime, at least sync_grains is a working workaround to make all the grains update properly with no flashbacks)

**Comment by markuskramerIgitt** [comment]
I now see a huge influence of “refresh“ in “sync_grains”. 
I have reproduced the steps below on a single minion many times with always the same result.
I retried within seconds or minutes, the result is always the same.

**Summary:**
The workaround only works when ```refresh:True``` but with the (false) statement ```No minions matched```.



Master is 2015.5.3 Debian 7, minion is 2015.5.3 Windows 7

----------------------This is state “seedSync.sls” --------------------------

```
sync dir:
  file.recurse:
    - name: /DATA
    - clean: True
    - source: salt:/DATA

repair grains flashbacks:
  module.run:
    - name: saltutil.sync_grains
    - refresh: False

```
----------------------This is state “seedSync.sls” --------------------------
I first set   ```refresh: False``` ,  then   ```refresh: True```

The minion has a grain ```role:seed```.

When I issue ```sudo salt -G 'role:seed' state.sls seedSync``` with ```refresh:False```,  salt replies ```No minions matched the target. No command was sent, no jid was assigned.``` but (surprisingly) executes the state exactly one time.  
After that, grain matching is broken, meaning that I only get ```No minions matched``` .

I can repair grain matching by addressing the minion with its name and syncing the grains (```sudo salt SEED1 saltutil.sync_grains```)

When I issue ```sudo salt -G 'role:seed' state.sls seedSync``` with ```refresh:True```, the behavior is exactly as above (execution with false ```No minion matched``` statement) ,  but I can issue the command  many times  over - grain matching continues to work (despite the statement) .
Other grain matching commands (e.g. ```sudo salt -G 'role:seed' test.ping```) work nicely without the ```No minions matched``` statement.

```/var/log/salt/master``` contains no entry of that time 

Could my Salt installation be broken in such a fundamental way?
Is grain matching unreliable for Windows minions? 
Is it a clue that this is completely reproducible?

I also have the suspicion that I use file.recurse in some “wrong way”: after I copied files to the salt file server (```/srv/salt```), do I have to perform a sanitary action?

I would like to gain as much understanding as possible, before re-installing or upgrading.

**Comment by basepi** [comment]
That's good information. It's possible there's something weird going on with your setup, I am very surprised that you're getting the `No minions matched` statement but the minion still seems to be getting the job. But I also didn't realize you were on a windows minion, which may also be part of the problem. (All windows, or just some?)
